### ISSUE TITLE
False positive `protected-access` and `unused-private-member` with Singleton Pattern

### ISSUE THREAD

**Issue by danilobecke** [pylint-dev/pylint#8987]
### Bug description

```python
from __future__ import annotations
from tomllib import load
from typing import Optional, Any

class Config:
    __shared: Optional[Config] = None

    def __init__(self, config_path: str) -> None:
        with open(config_path, 'rb') as config_file:
            self.__dict = load(config_file)

    @staticmethod
    def get(key_path: str) -> Optional[Any]:
        keys = key_path.split('.')
        keys_size = len(keys)
        if not Config.__shared:
            return None
        current_dict = Config.__shared.__dict.copy()
        value: Any | None = None
        for position, key in enumerate(keys):
            if position == keys_size - 1:
                value = current_dict[key] # last element
                break
            current_dict = dict[str, Any](current_dict[key])
        return value

    @staticmethod
    def initialize(config_path: str) -> None:
        if Config.__shared is not None:
            return
        Config.__shared = Config(config_path)
```

### Command used

```shell
pylint code/helpers/config.py
```

### Pylint output

```shell
************* Module helpers.config
code/helpers/config.py:18:23: W0212: Access to a protected member __dict of a client class (protected-access)
code/helpers/config.py:10:12: W0238: Unused private member `Config.__dict` (unused-private-member)
```

### Expected behavior

No issues

### Pylint version

```shell
pylint==2.17.4
astroid==2.15.5
Python 3.11.4
```

### OS / Environment

macOS
